backend/backend_cavity.py

```diff
@@ -0,0 +1,159 @@
+from collections import OrderedDict
+from datetime import datetime
+from typing import Dict
+
+from epics import caput
+from backend.fault import Fault, FaultCounter, PVInvalidError
+
+
+from lcls_tools.superconducting.sc_linac import Cavity
+from utils.utils import (
+    STATUS_SUFFIX,
+    DESCRIPTION_SUFFIX,
+    SEVERITY_SUFFIX,
+    parse_csv,
+    SpreadsheetError,
+    display_hash,
+)
+
+
+class BackendCavity(Cavity):
+    def __init__(
+        self,
+        cavity_num,
+        rack_object,
+    ):
+        super(BackendCavity, self).__init__(
+            cavity_num=cavity_num, rack_object=rack_object
+        )
+        self.status_pv: str = self.pv_addr(STATUS_SUFFIX)
+        self.severity_pv: str = self.pv_addr(SEVERITY_SUFFIX)
+        self.description_pv: str = self.pv_addr(DESCRIPTION_SUFFIX)
+
+        self.faults: OrderedDict[int, Fault] = OrderedDict()
+        self.create_faults()
+
+    def create_faults(self):
+        for csv_fault_dict in parse_csv():
+            level = csv_fault_dict["Level"]
+            suffix = csv_fault_dict["PV Suffix"]
+            rack = csv_fault_dict["Rack"]
+
+            if level == "RACK":
+                # Rack A cavities don't care about faults for Rack B and vice versa
+                if rack != self.rack.rack_name:
+                    # Takes us to the next iteration of the for loop
+                    continue
+
+                # tested in the python console that strings without one of these
+                # formatting keys just ignores them and moves on
+                prefix = csv_fault_dict["PV Prefix"].format(
+                    LINAC=self.linac.name,
+                    CRYOMODULE=self.cryomodule.name,
+                    RACK=self.rack.rack_name,
+                    CAVITY=self.number,
+                )
+                pv = prefix + suffix
+
+            elif level == "CRYO":
+                prefix = csv_fault_dict["PV Prefix"].format(
+                    CRYOMODULE=self.cryomodule.name, CAVITY=self.number
+                )
+                pv = prefix + suffix
+
+            elif level == "SSA":
+                pv = self.ssa.pv_addr(suffix)
+
+            elif level == "CAV":
+                pv = self.pv_addr(suffix)
+
+            elif level == "CM":
+                cm_type = csv_fault_dict["CM Type"]
+                prefix = csv_fault_dict["PV Prefix"].format(
+                    LINAC=self.linac.name,
+                    CRYOMODULE=self.cryomodule.name,
+                    CAVITY=self.number,
+                )
+
+                if (cm_type == "1.3" and self.cryomodule.is_harmonic_linearizer) or (
+                    cm_type == "3.9" and not self.cryomodule.is_harmonic_linearizer
```

Comment:
Is cm_type == "3.9" a possibility? I only see "1.3" or "ALL" or nothing appear in the CM Type column of the spreadsheet. Or is this here for readability and/or to make it easier to implement stuff in the future?

Reply:
Yeah it was for future proofing

Reply:
Ah okay, makes sense:)